File: Interview/house-robber.java
//  House Robber

class Solution {
    public int rob(int[] nums) {
        int size = nums.length;
        if (size <= 0)  return 0;
        if (size == 1)  return nums[0];
        if (size == 2)  return (nums[0]>nums[1] ? nums[0] : nums[1]);
        nums[1] = nums[0]>nums[1] ? nums[0] : nums[1];
        for (int i=2; i<size; i++)
            nums[i] = Math.max(nums[i-2] + nums[i], nums[i-1]);
        return nums[size-1];
    }
}

File: leetcode/house-robber.cc
//  House Robber

class Solution {
public:
    int rob(vector<int>& nums) {
        size_t n = nums.size();
        if (n == 0) return 0;
        if (n == 1) return nums[0];
        if (n == 2) return (nums[0]>nums[1] ? nums[0] : nums[1]);
        int pp = nums[0], p = (nums[0]>nums[1] ? nums[0] : nums[1]);
        int res;
        for (size_t i = 2; i < n; ++ i) {
            res = max(pp + nums[i], p);
            pp = p;
            p = res;
        }
        return res;
    }
};